### ISSUE TITLE
ValueError: Given `time_step`: TimeStep

### ISSUE THREAD

**Issue by Shinwazu** [tensorflow/agents#766]
```
import random
from abc import ABC
from random import choice

import numpy as np
from tf_agents.environments import py_environment
from tf_agents.specs import array_spec

from tf_agents.trajectories import time_step as ts, time_step, time_step_spec
from tf_agents.environments import utils


class WarehouseEnv(py_environment.PyEnvironment, ABC):

    def __init__(self, row, col):
        super(WarehouseEnv).__init__()
        self._action_spec = array_spec.ArraySpec(shape=(), dtype=np.int32, name='action')
        print(self._action_spec)
        self._observation_spec = array_spec.ArraySpec(shape=(9, 9), dtype=np.int32, name='observation')
        print(self._observation_spec)
        self._reward_spec = array_spec.ArraySpec(shape=(), dtype=np.int32, name='reward')
        print(self._reward_spec)
        self.ARTICLE_DICT = dict(Cola=1, Sprite=2, Fanta=3, Cola_Zero=4, Cola_Light=5, Mezzo_Mix=6)
        print(time_step_spec)
        # (ARTICLE_DICT)
        self.ARTICLE = list(self.ARTICLE_DICT.values())
        # print(ARTICLE)

        self.DEMAND_WAREHOUSE = [['1', '2', '3', '1', '1', '1'],
                                 ['1', '2'],
                                 ['3', '2', '1', '1'],
                                 ['4'],
                                 ['1', '6', '1']]

        self._episode_ended = False
        self.row = row
        self.col = col
        self.Supply_Warehouse = np.zeros([self.row, self.col], dtype=np.int32)
        print(self.Supply_Warehouse)
        self._state = self.Supply_Warehouse

    def action_spec(self):
        return self._action_spec

    def observation_spec(self):
        return self._observation_spec

    def _reset(self):
        self._state = self.Supply_Warehouse
        self._episode_ended = False
        return time_step.restart(np.array(self._state, dtype=np.int32))

    def add_to_supply_warehouse(self):
        for row in range(self.row):
            for col in range(self.col):
                if self.Supply_Warehouse[row][col] == 0:
                    if self.Supply_Warehouse[row - 1][col - 1] != 0:
                        self.Supply_Warehouse[row][col] += random.choice(list(self.ARTICLE_DICT.values()))
                        break
                return self._state

    def check_demand(self):
        for row in self._state:
            for col in self._state:
                if self._state[row][col] > self._state[:, 1] or self._state[row][col] < self._state[row + 1, col + 1] == 0:
                    if self._state[row][col] == self.DEMAND_WAREHOUSE[:, :]:
                        return self._state and self.DEMAND_WAREHOUSE

    def remove_article(self):
        for row in range(self.row):
            for col in range(self.col):
                if self.check_demand is True:
                    self.DEMAND_WAREHOUSE[:, :] = 0
                    return self.Supply_Warehouse

    def _step(self, action):
        if self._episode_ended:
            return self.reset()
        action = action.item()
        if action == 0:  # add article to the supply warehouse
            self.add_to_supply_warehouse()
            new_article = random.choice(list(self.ARTICLE_DICT.values()))
            self._state += new_article
            return ts.transition(np.array(self._state, dtype=np.int32), reward=1, discount=[1])
        if all in self._state > 0:
            self._episode_ended = True
            return ts.termination(np.array(self._state, dtype=np.int32), reward=-9)

        if action == 1:  # Checks demand arbitrarily and remove article from supply warehouse and similar value in
            # Demand list becomes 0
            check = [self.add_to_supply_warehouse, self.check_demand]
            choice(check)()
            if self.check_demand is True:
                self.remove_article()
                return ts.transition(np.array(self._state, dtype=np.int32), reward=5, discount=[1])
            elif len(self.DEMAND_WAREHOUSE) == 0:
                self._episode_ended = True
                return ts.termination(np.array(self._state, dtype=np.int32), reward=2)
            else:
                return ts.transition(np.array(self._state, dtype=np.int32), reward=0, discount=[1])

        info = {}
        return ts.termination(np.array(self._state, dtype=np.int32), reward=2), info

```
2. 
```
from tf_agents.environments import tf_py_environment, utils
from WarehouseEnv import WarehouseEnv

python_environment = WarehouseEnv(9, 9)
utils.validate_py_environment(python_environment, episodes=5)
tf_env = tf_py_environment.TFPyEnvironment(python_environment)
```

Description:
I'm trying to design and simulate a warehouse environment where adding the article to the warehouse (column wise if the previous column is filled).
Checking demand list and if the article in the demand list (article from '0' index in demand list) matches the articles in warehouse (articles in columns towards the right side) then both values = 0 (i.e., article from warehouse is removed to satisfy demand and hence demand is satisfied, the similar value in demand list is also cleared).
This function after remmoving the article is called removing article.
This cycle contnuous until demand list is 0 or warehouse is full.
rewards are based on how I want the Env to behave.

But I have a value error if I use bounded (Discrete )array:

ValueError: Given "time_step: TimeStep(

{° discount': array(1., dtype=fLoat32),

"observation': array(IL[0, 0, 0, 0, 0],

[0, 0, 0, 0, 0],

[0, 0, 0, 0, 0],

[o,

0, 0, 0, 0],

[0, 0, 0, 0, 0]]], dtype=int32),

*reward': array(0., dtype=FLoat32),

'step_type": array(0, dtype=int32)}) does not match expected 'time_step_spec': Timestep(

{'discount': BoundedArraySpec (shape=), dtype=dtype('FLoat32'), name='discount', minimum=0.0, maximum=1.0),

'observation': BoundedArraySpec (shape=(5, 5), type=dtype ('int32'), name='observation', minimum=0, maximum=10),

"reward': ArraySpec (shape=), dtype=dtype ('FLoat32'), name='reward'),

"step_type': ArraySpec(shape=), dtype=dtype('int32'), name='step_type')})

I tried it by changing the action to continuous which gives me this error:

line 29, in __init__
    self._action_spec = is_continuous(spec=np.float64).BoundedArraySpec(shape=(1,), dtype=np.float64, minimum=0,
AttributeError: 'bool' object has no attribute 'BoundedArraySpec'

I'm in desperate need of help. Please help!!!

**Comment by philstahlfeld** [comment]
It looks like your observation in that TimeStep has shape (1, 5, 5) based on the three closing square brackets, but your observation spec wants shape (5, 5). Can you try either removing the extra dimension on your observation or updating your observation_spec to be (1, 5, 5)?

**Comment by Shinwazu** [comment]
I'll give it a try.

**Comment by Shinwazu** [comment]
hello Phil, I have changed the action spec and observation spec to discrete and the error looks different now. I have provided a Colab link below. Would it be possible for you to look into it. If you use PyCharm, I can also give the collaboration code for Pycharm.

https://colab.research.google.com/drive/1nLLt1QwefHL3t9qgpNJt2lhIVZmMuplt?usp=sharing

**Comment by philstahlfeld** [comment]
If you're creating a PyEnvironment, you should be using ArraySpecs/BoundedArraySpecs instead of TensorSpecs. 

The mismatch you're seeing when running utils.validate_py_environment is in the shape of the observation_spec and the returned observation. The observation spec dictates a shape of (9, 9), but all of the returns in `_step` have a shape of (1, 9, 9). The reason for this is that `self._state` has shape (9, 9) so calling `np.array([self._state])` will have shape (1, 9, 9). I think what you want to do is just call `np.array(self._state)` (without the square brackets).

**Comment by Shinwazu** [comment]
Thanks a lot.

the observation space is being printed but the action spec is not iterating.
is there a way to check the observation spec being updated after every action or to check if its taking a action

[[0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0]
 [0 0 0 0 0 0 0 0 0]]

I have corrected the mistakes [] and when I execute it, I get the following error;

Traceback (most recent call last):
  File "/Users/shinwazu/PycharmProjects/pythonProject/WarehouseEnv(Tensorflow)/TensorflowEnv.py", line 5, in <module>
    utils.validate_py_environment(python_environment, episodes=5)
  File "/Users/shinwazu/.local/lib/python3.10/site-packages/tf_agents/environments/utils.py", line 85, in validate_py_environment
    episode_count += np.sum(time_step.is_last())
AttributeError: 'tuple' object has no attribute 'is_last'

**Comment by philstahlfeld** [comment]
That seems to be from your last return in `_step`, which is `return self._observation_spec, 0, info`. You need to return a TimeStep like `ts.transition` or `ts.termination`.

**Comment by Shinwazu** [comment]
`return ts.termination(np.array(self._state, dtype=np.int32), reward=2), info`

Its still executing with the same error. I'm returning a TimeStep as above instead of `return self._observation_spec, 0, info`

**Comment by philstahlfeld** [comment]
You can't return `info`. It has to just be `return ts.termination(np.array(self._state, dtype=np.int32), reward=2)`.

**Comment by Shinwazu** [comment]
I'm getting the following error;

Exception ignored in: <function Pool.__del__ at 0x137c0f7f0>
Traceback (most recent call last):
  File "/Users/shinwazu/miniforge3/envs/tensorflow/lib/python3.10/multiprocessing/pool.py", line 268, in __del__
  File "/Users/shinwazu/miniforge3/envs/tensorflow/lib/python3.10/multiprocessing/queues.py", line 371, in put
AttributeError: 'NoneType' object has no attribute 'dumps'


I don't understand 'NoneType' object. Is it from my obs spec

**Comment by philstahlfeld** [comment]
I was able to run `utils.validate_py_environment` on your environment for 10k steps and it didn't error out so I think your environment is good to go. It sounds like the error you're running into now is a different error (perhaps in your training pipeline). Since we've worked through the error related to this issue, I would suggest closing this issue. If you're still encountering the error in your previous comment after doing some more debugging, I would suggest creating a new issue (ideally, with more context that would help others debug).
